content/advent-2016/machine-learning-google-prediction-api.md

```diff
@@ -0,0 +1,224 @@
+
+Creating training data isn't easy, and is often the most delecate part of the process.
+Luckily Google has provided us with a set of data that we can use.
+
+The Language Identification dataset attempts to train a model that will allow the machine to predict
+which language a given sentence is, either English, Spanish, or French.
+
+The data is in CSV format, and looks something like this:
+
+```
+"English", "This version of the simple language detection model was created on December 27, 2011..."
+"French", "M. de Troisvilles, comme s'appelait encore sa famille en Gascogne, ou M. de Tréville..."
+"Spanish", "En efeto, rematado ya su juicio, vino a dar en el más estraño pensamiento que..."
+...
+```
+
+* Have a quick [look at the entire dataset](https://cloud.google.com/prediction/docs/language_id.txt)
+
+As you can see, there are hundreds of examples of each language. The first column is the answer, and the
+second column is an example of a question that we might ask it.
+
+Google has already trained a model with this data, so we don't need to do it - but it's as simple as creating
+a Google Project, uploading a CSV file to Google Cloud Storage, and making an API request similar to:
+
+```
+POST https://www.googleapis.com/prediction/v1.6/projects/{project-id}/trainedmodels
+{
+    "id": "model-identifier",
+    "storageDataLocation": "path/to/training-data.csv"
+}
+```
+
+Once the model is ready, we can start to ask it to categorise sentences by making requests similar to:
+```
+POST https://www.googleapis.com/prediction/v1.6/projects/{project-id}/trainedmodels/model-identifier/predict
+{
+    "input": {
+        "csvInstance": [
+            "Please predict which language this is"
+        ]
+    }
+}
+```
+
+Yes, it is a little weird that we're embedding CSV into JSON, and I expect this will evolve in future versions
+of the API. But it is still pretty clear what's going on; we're asking the Google Prediction API to predict which
+language the `"Please predict which language this is"` sentence is in.
+
+## Building a little app in Go
+
+We will write a little command line tool that lets us query the model from the terminal. 
+
+As the example project for this article was being put together, it became clear that the lion share of the work
+was actually in authorising the requests via OAuth. And the experience for a terminal app was less than desirable; you 
+had to copy and paste a URL into a browser, log in (or create an account), click Accept to allow access to the API, 
+then copy the access code from the query parameters of a failed redirect and paste it into the terminal.
+
+OAuth is a web based authorisation protocol through and through, and short of implementing an HTTP endpoint to
+receive the request, there isn't much we can do to make the experience any better. So instead, we are going to be
+a little cheeky and piggy back on [Google's Try it app hosted on App Engine](http://try-prediction.appspot.com).
```

Comment:
Why not start an HTTP server locally, show the URL to the server which triggers OAuth2's implicit flow to get the access token, saves the final access token and stops the server?